x-pack/plugins/ml/public/application/components/ml_inference/components/test_pipeline.tsx

```diff
@@ -95,36 +124,79 @@ export const TestPipeline: FC<Props> = memo(({ state, sourceIndex }) => {
     setIsValid(valid);
   };
 
-  const getSampleDocs = useCallback(async () => {
-    let records: IngestSimulateDocument[] = [];
-    let resp;
+  const getDocs = useCallback(
+    async (body: any) => {
+      let records: IngestSimulateDocument[] = [];
+      let resp;
+      try {
+        resp = await esSearch(body);
 
-    try {
-      resp = await esSearch({
-        index: sourceIndex,
-        body: {
-          size: 1,
-        },
-      });
+        if (resp && resp.hits.total.value > 0) {
+          records = resp.hits.hits;
+        }
+      } catch (error) {
+        // eslint-disable-next-line no-console
+        console.error(error);
+      }
+      setSampleDocsString(JSON.stringify(records, null, 2));
+      setSimulatePipelineResult(undefined);
+      setLastFetchedSampleDocsString(JSON.stringify(records, null, 2));
+      setIsValid(true);
+    },
+    [esSearch]
+  );
+
+  const { getSampleDoc, getRandomSampleDoc } = useMemo(
+    () => ({
+      getSampleDoc: async () => {
+        getDocs({
+          index: sourceIndex,
+          body: {
+            size: 1,
+          },
+        });
+      },
+      getRandomSampleDoc: async () => {
+        getDocs({
+          index: sourceIndex,
+          body: {
+            size: 1,
+            query: {
+              function_score: {
+                query: { match_all: {} },
+                random_score: {},
+              },
+            },
+          },
+        });
+      },
+    }),
+    [getDocs, sourceIndex]
```

Comment:
If `getDocs` is nowhere else used than in this `useMemo`, I was hoping we could move it in there to have just one hook being used for all these queries, what do you think?

Reply:
That's true, although they have different dependencies. As is, `getSampleDoc` and `getRandomizedSampleDoc` functions are only created once and `getDocs` only changes as the arguments change. 
I might be missing some context but it's unclear to me what the benefits of having it all in one hook are?

Reply:
It's just that multiple hooks that depend on each other might cause multiple re-renders, whereas all code in a single `useMemo/useCallback` will just cause one rerender.